### PR TITLE
Properties support for @ConfigProperty on method parameters

### LINKED ISSUE #135: MicroProfile ConfigProperty annotations on methods not being processed
@ConfigProperty can be used also on methods so we should analyze this binding as well

### PR DESCRIPTION
**Pull request by xorye**
Fixes #135 

To test this PR, please use the `config-quickstart` test project. In that project, there will be ConfigProperty properties coming from constructor parameters and regular method (non-constructor) parameters.

Demo:
demo

### CODE REVIEW

File: microprofile.jdt/com.redhat.microprofile.jdt.core/src/main/java/com/redhat/microprofile/jdt/core/AbstractAnnotationTypeReferencePropertiesProvider.java [hunk 1]
```diff
@@ -51,29 +54,51 @@ protected SearchPattern createSearchPattern(String annotationName) {
...
+				LOGGER.log(Level.SEVERE,
+						"Cannot compute MicroProfile properties for the Java element '" + javaElement != null
+								? javaElement.getElementName()
+								: match.getElement() + "'.",
+						e);
+			}
+		}
+	}
+
+	protected void processAnnotation(IJavaElement javaElement, SearchContext context, IProgressMonitor monitor)
```
Comment (angelozerr) [inline microprofile.jdt/com.redhat.microprofile.jdt.core/src/main/java/com/redhat/microprofile/jdt/core/AbstractAnnotationTypeReferencePropertiesProvider.java]: Add javadoc

File: microprofile.jdt/com.redhat.microprofile.jdt.core/src/main/java/com/redhat/microprofile/jdt/core/AbstractAnnotationTypeReferencePropertiesProvider.java [hunk 2]
```diff
@@ -51,29 +54,51 @@ protected SearchPattern createSearchPattern(String annotationName) {
...
+			throws JavaModelException {
+		IAnnotation[] annotations = ((IAnnotatable) javaElement).getAnnotations();
+		for (IAnnotation annotation : annotations) {
+			processAnnotation(javaElement, context, monitor, annotation);
+		}
+	}
+
+	private void processAnnotation(IJavaElement javaElement, SearchContext context, IProgressMonitor monitor,
```
Comment (angelozerr) [inline microprofile.jdt/com.redhat.microprofile.jdt.core/src/main/java/com/redhat/microprofile/jdt/core/AbstractAnnotationTypeReferencePropertiesProvider.java]: Add javadoc

File: microprofile.jdt/com.redhat.microprofile.jdt.core/src/main/java/com/redhat/microprofile/jdt/core/PropertiesManager.java [hunk 3]
```diff
@@ -509,10 +510,13 @@ public IMember findDeclaredProperty(IJavaProject javaProject, String sourceType,
 				// Method signature has been generated with JDT API, so we are sure that we have
 				// a ')' character.
 				int endBracketIndex = sourceMethod.indexOf(')');
-				String methodSignature = sourceMethod.substring(startBracketIndex + 1, endBracketIndex);
+				String methodSignature = sourceMethod.substring(startBracketIndex, endBracketIndex + 1);
```
Comment (angelozerr) [inline microprofile.jdt/com.redhat.microprofile.jdt.core/src/main/java/com/redhat/microprofile/jdt/core/PropertiesManager.java]: why did you change this code?

File: microprofile.jdt/com.redhat.microprofile.jdt.core/src/main/java/com/redhat/microprofile/jdt/core/AbstractAnnotationTypeReferencePropertiesProvider.java [hunk 4]
```diff
@@ -51,29 +54,51 @@ protected SearchPattern createSearchPattern(String annotationName) {
...
-			String[] names = getAnnotationNames();
-			IAnnotation[] annotations = ((IAnnotatable) javaElement).getAnnotations();
-			for (IAnnotation annotation : annotations) {
-				for (String annotationName : names) {
-					if (isMatchAnnotation(annotation, annotationName)) {
-						processAnnotation(javaElement, annotation, annotationName, context, monitor);
-						break;
-					}
+			Object element = match.getElement();
+			if (match instanceof TypeReferenceMatch) {
```
Comment (xorye) [inline microprofile.jdt/com.redhat.microprofile.jdt.core/src/main/java/com/redhat/microprofile/jdt/core/AbstractAnnotationTypeReferencePropertiesProvider.java]: I noticed that in the case where the match is from a method parameter or a Java field, they both cause the `match` instance to be a  `TypeReferenceMatch`. 

I noticed the following: 
If it is from a method parameter, localElement exists
If it is from a Java field, localElement does not exist

File: microprofile.jdt/com.redhat.microprofile.jdt.core/src/main/java/com/redhat/microprofile/jdt/core/utils/JDTTypeUtils.java [hunk 5]
```diff
@@ -42,6 +44,27 @@ public static IType findType(IJavaProject project, String name) {
 		}
 	}
 
+	public static String getResolvedTypeName(IJavaElement javaElement) {
```
Comment (angelozerr) [inline microprofile.jdt/com.redhat.microprofile.jdt.core/src/main/java/com/redhat/microprofile/jdt/core/utils/JDTTypeUtils.java]: Add javadoc

File: microprofile.jdt/com.redhat.microprofile.jdt.core/src/main/java/com/redhat/microprofile/jdt/core/utils/JDTTypeUtils.java [hunk 6]
```diff
@@ -42,6 +44,27 @@ public static IType findType(IJavaProject project, String name) {
...
+		case IJavaElement.LOCAL_VARIABLE:
+			return getResolvedTypeName((ILocalVariable) javaElement);
+		case IJavaElement.FIELD:
+			return getResolvedTypeName((IField) javaElement);
+		default:
+			return null;
+		}
+	}
+
+	public static String getResolvedTypeName(ILocalVariable localVar) {
```
Comment (angelozerr) [inline microprofile.jdt/com.redhat.microprofile.jdt.core/src/main/java/com/redhat/microprofile/jdt/core/utils/JDTTypeUtils.java]: Add javadoc

File: microprofile.jdt/com.redhat.microprofile.jdt.core/src/main/java/com/redhat/microprofile/jdt/core/utils/JDTTypeUtils.java [hunk 7]
```diff
@@ -97,11 +120,37 @@ public static String getPropertyType(IType type, String typeName) {
 		return type != null ? type.getFullyQualifiedName('.') : typeName;
 	}
 
+	public static boolean isBinary(IJavaElement javaElement) {
```
Comment (angelozerr) [inline microprofile.jdt/com.redhat.microprofile.jdt.core/src/main/java/com/redhat/microprofile/jdt/core/utils/JDTTypeUtils.java]: Add javadoc

File: microprofile.jdt/com.redhat.microprofile.jdt.core/src/main/java/com/redhat/microprofile/jdt/core/utils/JDTTypeUtils.java [hunk 8]
```diff
@@ -97,11 +120,37 @@ public static String getPropertyType(IType type, String typeName) {
...
+		if (javaElement instanceof IMember) {
+			return ((IMember) javaElement).isBinary();
+		} else if (javaElement instanceof ILocalVariable) {
+			return isBinary(((ILocalVariable) javaElement).getDeclaringMember());
+		}
+		return false;
+	}
+
+	public static String getSourceType(IJavaElement javaElement) {
```
Comment (angelozerr) [inline microprofile.jdt/com.redhat.microprofile.jdt.core/src/main/java/com/redhat/microprofile/jdt/core/utils/JDTTypeUtils.java]: Add javadoc

File: microprofile.jdt/com.redhat.microprofile.jdt.core/src/main/java/com/redhat/microprofile/jdt/core/utils/JDTTypeUtils.java [hunk 9]
```diff
@@ -97,11 +120,37 @@ public static String getPropertyType(IType type, String typeName) {
...
+		case IJavaElement.FIELD:
+			return getSourceType((IField) javaElement);
+		case IJavaElement.METHOD:
+			return getSourceType((IMethod) javaElement);
+		default:
+			return null;
+		}
+	}
+
+	public static String getSourceType(ILocalVariable member) {
```
Comment (angelozerr) [inline microprofile.jdt/com.redhat.microprofile.jdt.core/src/main/java/com/redhat/microprofile/jdt/core/utils/JDTTypeUtils.java]: Add javadoc

File: microprofile.jdt/com.redhat.microprofile.jdt.core/src/main/java/com/redhat/microprofile/jdt/core/utils/JDTTypeUtils.java [hunk 10]
```diff
@@ -97,11 +120,37 @@ public static String getPropertyType(IType type, String typeName) {
...
+		return getSourceType(member.getDeclaringMember());
+	}
+
 	public static String getSourceType(IMember member) {
 		return getPropertyType(member.getDeclaringType(), null);
 	}
 
-	public static String getSourceField(IField field) {
+	public static String getSourceField(IJavaElement field) {
```
Comment (angelozerr) [inline microprofile.jdt/com.redhat.microprofile.jdt.core/src/main/java/com/redhat/microprofile/jdt/core/utils/JDTTypeUtils.java]: Add javadoc

File: microprofile.jdt/com.redhat.microprofile.jdt.core/src/main/java/com/redhat/microprofile/jdt/core/PropertiesManager.java (hunk 3 above)
Comment (xorye) [reply microprofile.jdt/com.redhat.microprofile.jdt.core/src/main/java/com/redhat/microprofile/jdt/core/PropertiesManager.java]: For calling `Signature.getParameterTypes(methodSignature)`, the method signature needed the `(` and `)` around it. I changed this code because previously, the brackets were not being included in the substring

Comment (xorye) [reply microprofile.jdt/com.redhat.microprofile.jdt.core/src/main/java/com/redhat/microprofile/jdt/core/PropertiesManager.java]: It was for go to definiton to work

File: microprofile.jdt/com.redhat.microprofile.jdt.core/src/main/java/com/redhat/microprofile/jdt/core/AbstractAnnotationTypeReferencePropertiesProvider.java [hunk 11]
```diff
@@ -51,29 +53,70 @@ protected SearchPattern createSearchPattern(String annotationName) {
...
+				Object localElement = ((TypeReferenceMatch) match).getLocalElement();
+				element = localElement != null ? localElement : element;
+			}
+			
+			if (element instanceof IAnnotation) {
+				// ex : for Local variable
+				IAnnotation annotation = ((IAnnotation) element);
+				javaElement = annotation.getParent();
+				processAnnotation(javaElement, context, monitor, annotation);
+			} else {
```
Comment (angelozerr) [inline microprofile.jdt/com.redhat.microprofile.jdt.core/src/main/java/com/redhat/microprofile/jdt/core/AbstractAnnotationTypeReferencePropertiesProvider.java]: write direct `} else if (`

File: microprofile.jdt/com.redhat.microprofile.jdt.core/src/main/java/com/redhat/microprofile/jdt/core/AbstractAnnotationTypeReferencePropertiesProvider.java (hunk 4 above)
Comment (angelozerr) [reply microprofile.jdt/com.redhat.microprofile.jdt.core/src/main/java/com/redhat/microprofile/jdt/core/AbstractAnnotationTypeReferencePropertiesProvider.java]: create a private static class method Object getMatchedElement(SearchMatch match) which does that and add comments that you explained below.

Comment (xorye) [reply microprofile.jdt/com.redhat.microprofile.jdt.core/src/main/java/com/redhat/microprofile/jdt/core/AbstractAnnotationTypeReferencePropertiesProvider.java]: fixed

**Comment by xorye** [comment]
In my latest push, I have addressed the PR feedback and added two tests in `PropertiesManagerLocationTest.java`

**Comment by xorye** [comment]
This PR has two commits, should I change the commit message of the older commit? Maybe I should just squash them to one commit?

**Comment by angelozerr** [comment]
Yes one commit is enough

**Comment by angelozerr** [comment]
The properties support in application.properties works pretty well. Could you create an issue to improve the find definition to jump to the parameter (and not not to the constructor)?

The hover in java files doesn't work (to display the value defined in application.properties). please create an issue for that (or if it's easy to do, fix it in this PR).

**Comment by xorye** [comment]
It has been changed to one commit

**Comment by angelozerr** [comment]
Thanks @xorye !
